### PR TITLE
Share data via chat

### PR DESCRIPTION
**Pull request by Wooza**
I had a go at #21. This branch is based on #23. The prospecting data can be shared via `.pi share`. The major downside of course is that we abuse the global chat as a broadcast channel. I also noticed that the client struggles to print the huge base64 encoded string, i.e., the client is unresponsive for a few seconds. This problem seems to persist afterwards when using the chat, which is kind of problematic, because this breaks the chat for everyone. 
A wild guess of mine is that the text renderer struggles because it tries to apply word wrapping but the string has no whitespace.

Edit: When adding regular whitespace to the base64 string, the lag spikes no longer occur.

### CODE REVIEW

File: src/Utils/ChatDataSharing.cs [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+        private readonly ICoreClientAPI api;
+        private readonly ProspectorOverlayLayer prospectorOverlayLayer;
+        public ChatDataSharing(ICoreClientAPI api, ProspectorOverlayLayer prospectorOverlayLayer)
+        {
+            this.api = api;
+            this.prospectorOverlayLayer = prospectorOverlayLayer;
+            api.Event.ChatMessage += OnChatMessage;
+        }
+
+        private void OnChatMessage(int groupId, string message, EnumChatType chattype, string data) {
```
Comment (p3t3rix) [inline src/Utils/ChatDataSharing.cs:28]: I wonder if we need to make this more opt-in, as a malicious user could invalidate/destroy the data a user has collected.
I think at least a settings that needs to be activated before the data will be blindly accepted should be added.
Other ideas are time/message limited activation (like you need to press a button to accept the next messages/X minutes)
Or an extra dialog you need to confirm (and maybe optional if your data should only be expanded).

File: src/Utils/ChatDataSharing.cs [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+                stringBuilder.Append(input.Substring(currentPosition));
+            }
+            return stringBuilder.ToString();
+        }
+
+        private T DeserializeFromBase64<T>(string message) where T : class 
+        {
+            try
+            {
+                using (var resultStream = new MemoryStream(Convert.FromBase64String(message)))
```
Comment (p3t3rix) [inline src/Utils/ChatDataSharing.cs:102]: good idea to compress the whole stuff beforehand 👍

File: src/Utils/ChatDataSharing.cs [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+        public void ShareData(ProspectorMessages messages)
+        {
+            string data = SerializeToBase64<ProspectorMessages>(messages);
+            if (data == null) 
+            {
+                return;
+            }
+            // Add some whitespace before sending, to avoid lag spikes when rendering the chat message.
+            // The base64 deserializer ignores whitespace.
+            api.SendChatMessage(CHAT_PREFIX + AddWhitespace(data, 64));
```
Comment (p3t3rix) [inline src/Utils/ChatDataSharing.cs:56]: Have you notified @tyronx about the issue ?

Comment (Wooza) [reply src/Utils/ChatDataSharing.cs:56]: Not yet. Tbh, I'm a bit ashamed of this solution, as the chat is clearly not intended for such a use case. 

I guess a better solution would be to change the client api to add some kind of broadcast, where the server just relays a packet to all clients. Clients that recognize the packet, i.e., have a certain handler from a mod, can then process the packet. I will open an issue when I find the time.

File: src/Utils/ChatDataSharing.cs (hunk 1 above)
Comment (Wooza) [reply src/Utils/ChatDataSharing.cs:28]: Fair point. Maybe one has to explicitly enable an option to receive updates from the chat? Alternatively, we could send the message in a certain group chat, so people have some setup to do before exchanging data.

File: src/Utils/ChatDataSharing.cs (hunk 3 above)
Comment (tyronx) [reply src/Utils/ChatDataSharing.cs:56]: I am aware of the text performance limits of the chat window, yes

File: src/Utils/ChatDataSharing.cs (hunk 1 above)
Comment (Wooza) [reply src/Utils/ChatDataSharing.cs:28]: I added an opt-in to accept chat data

**Comment by Wooza** [comment]
Can be reviewed after merging #23

**Comment by p3t3rix** [comment]
> Can be reviewed after merging #23

Could you merge master into the branch now that i merged #23  ?

**Comment by Wooza** [comment]
I guess you can close #25 and #28, as these should be fixed with #23

**Comment by p3t3rix** [comment]
Well i am not sure what exactly is wrong with my installation of VS, but i get JsonTesselator erros all over the place (the ones i thought would be related to the RebuildMap call, i guess i was wrong), so i can't really test the mod as much as i want :/
but since it seems to work i will just merge it and publish a release for 18.1.

**Pull request merged**